index.js

```diff
@@ -39,25 +53,60 @@ function walkSync (baseDir, _options, _relativePath) {
   if (m && !m.mayContain(relativePath)) {
     return results;
   }
+
   var entries = fs.readdirSync(baseDir + '/' + relativePath).sort();
   for (var i = 0; i < entries.length; i++) {
     var entryRelativePath = relativePath + entries[i];
-    var stats = getStat(baseDir + '/' + entryRelativePath);
+    var fullPath = baseDir + '/' + entryRelativePath;
+    var stats = getStat(fullPath);
 
     if (stats && stats.isDirectory()) {
       if (options.directories !== false && (!m || m.match(entryRelativePath))) {
-        results.push(entryRelativePath + '/');
+        results.push(new Entry(entryRelativePath + '/', baseDir, stats.mode, stats.size, stats.mtime.getTime()));
       }
-      results = results.concat(walkSync(baseDir, options, entryRelativePath));
+      results = results.concat(_walkSync(baseDir, options, entryRelativePath));
     } else {
       if (!m || m.match(entryRelativePath)) {
-        results.push(entryRelativePath);
+        results.push(new Entry(entryRelativePath, baseDir, stats && stats.mode, stats && stats.size, stats && stats.mtime));
       }
     }
   }
   return results;
 }
 
+function Entry(relativePath, basePath, mode, size, mtime) {
+  this.relativePath = relativePath;
+  this.basePath = basePath;
+  this.mode = mode;
+  this.size = size;
+  this.mtime = mtime;
+}
+
+Object.defineProperty(Entry.prototype, 'fullPath', {
+  get: function() {
+    return this.basePath + '/' + this.relativePath;
+  }
+});
+
+Entry.prototype.isEqual = function (other) {
+  return this.fullPath === other.fullPath &&
+         this.mode === other.mode &&
+         this.size === other.size &&
+         this.mtime === other.mtime;
+};
+
+Entry.prototype.isDirectory = function () {
+  return (this.mode & 61440) === 16384;
+};
+
+Entry.prototype.isFile = function () {
+  return (this.mode & 61440) === 32768;
```

Comment:
There are files that are neither directories nor regular files, such as sockets and FIFOs. We don't really support those, but should we group those under `isFile` for correctness' sake, so isFile is just `!isDirectory`?